PayloadPkg/OsLoader/OsLoader.c

```diff
@@ -628,13 +575,233 @@ StartBooting (
+    DEBUG ((DEBUG_ERROR, "Failed to init media - %r\n", Status));
+    return Status;
+  }
+
+  AddMeasurePoint (0x4050);
+  MediaTuning (BootMediumPciBase);
+  AddMeasurePoint (0x4055);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Find a MBR or GPT partition from the given hardware partition number
+
+  Based on given boot option, this function will find a MBR or GPT partition
+
+  @param[in]  OsBootOption      OS boot option to boot
+  @param[out] HwPartHandle      Hardware partition handle when detected successfully
+
+  @retval  EFI_SUCCESS          A hardware partition is initialized successfully
+  @retval  Others               An error when finding a partition
+
+**/
+EFI_STATUS
+EFIAPI
+FindBootPartitions (
+  IN  OS_BOOT_OPTION      *OsBootOption,
+  OUT EFI_HANDLE          *HwPartHandle
+  )
+{
+  EFI_STATUS                Status;
+  UINT8                     HwPart;
+
+  DEBUG ((DEBUG_INFO, "Try to find boot partition\n"));
+
+  ASSERT (OsBootOption != NULL);
+
+  HwPart = OsBootOption->HwPart;
+
+  Status = FindPartitions (HwPart, HwPartHandle);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to find partition - %r\n", Status));
+    ClosePartitions (*HwPartHandle);
+    return Status;
+  }
+  AddMeasurePoint (0x4060);
+
+  DEBUG ((DEBUG_INFO, "Find partition success\n"));
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Initialize File System
+
+  Based on given boot option, this function will initialize File System
+
+  @param[in]  OsBootOption      OS boot option to boot
+  @param[in]  HwPartHandle      Detected hardware partition handle
+  @param[out] FsHandle          File System handle when detected successfully
+
+  @retval  EFI_SUCCESS          A File System is initialized successfully
+  @retval  Others               An error when initializing a File System
+
+**/
+EFI_STATUS
+EFIAPI
+InitBootFileSystem (
+  IN  OS_BOOT_OPTION      *OsBootOption,
+  IN  EFI_HANDLE           HwPartHandle,
+  OUT EFI_HANDLE          *FsHandle
+  )
+{
+  EFI_STATUS      Status;
+  UINT32          SwPart;
+  UINT32          FsType;
+  INT32           BootSlot;
+
+  ASSERT (OsBootOption != NULL);
+
+  SwPart = OsBootOption->SwPart;
+  FsType = OsBootOption->FsType;
+
+  DEBUG ((DEBUG_INFO, "Init File system\n"));
+  if ((OS_FILE_SYSTEM_TYPE)FsType < EnumFileSystemMax) {
+    Status = InitFileSystem (SwPart, FsType, HwPartHandle, FsHandle);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_INFO, "No partitions found, Status = %r\n", Status));
+      return Status;
+    }
+  } else {
+    //
+    // Assume RAW format partition
+    //
+    DEBUG ((DEBUG_INFO, "Requested FsType %d. Assume RAW format partition.\n", FsType));
+    *FsHandle = NULL;
+
+    //
+    // Get boot image A/B info - Supported in RAW partition only.
+    //
+    BootSlot = GetBootSlot (OsBootOption, HwPartHandle);
+    DEBUG ((DEBUG_INFO, "BootSlot = 0x%x\n", BootSlot));
+    if (BootSlot == 1) {
+      OsBootOption->BootFlags |= LOAD_IMAGE_FROM_BACKUP;
+    }
```

Comment:
A/B support could be used for file system. e.g. Current TGL and EHL could boot Yocto (using Mender) from A or B partition.

Reply:
Got it. Let me update it in next commit.